### DISCUSSION TITLE
More hardware (display) support?

### DISCUSSION THREAD

**Discussion by edwinm** [microsoft/devicescript#393] [Q&A] [answered]
Is support for more hardware planned? Or is counted on support from contributors?

It would be very nice if DeviceScript would support cheap displays like SSD1306, ILI9341, ILI9225, 12864 and ST7735.

And of course many other hardware devices and chips.

**Comment by mmoskal** [comment, marked as answer]
Some things are easier to implement (simple GPIO, analog sensors, or I2C devices), while others are more complicated.

The display in particular needs SPI support (or that parallel protocol they use). Either the whole driver would need to be implemented in C or an SPI interface would need to be exposed to DeviceScript, probably directly and not through the Jacdac bus due to efficiency reasons.

We currently have no big plans for adding lots of hardware, but contributions are always welcome!

**Reply by mmoskal** [reply]
FYI I'm working on image APIs. SPI is in the esp32 main branch. There will be an image API to send it over SPI or i2c but we'll still need driver to set up the screen

**Reply by mmoskal** [reply]
Draft Image PR #423

**Reply by pelikhan** [reply]
The image support is in 2.11.1!

**Comment by pelikhan** [comment]
Tracked at https://github.com/microsoft/devicescript/issues/396
